service.go

```diff
@@ -101,26 +103,20 @@ type Opts struct {
 	GroupMessageStoreType  string
 }
 
-func (opts *Opts) applyPushDefaults() error {
+func (opts *Opts) applyPushDefaults() {
 	if opts.Logger == nil {
 		opts.Logger = zap.NewNop()
 	}
 
 	if opts.PrometheusRegister == nil {
 		opts.PrometheusRegister = prometheus.DefaultRegisterer
 	}
-
-	opts.applyDefaultsGetDatastore()
-
-	return nil
 }
 
 func (opts *Opts) applyDefaultsGetDatastore() {
 	if opts.RootDatastore == nil {
 		if opts.DatastoreDir == "" || opts.DatastoreDir == InMemoryDirectory {
 			opts.RootDatastore = ds_sync.MutexWrap(ds.NewMapDatastore())
-		} else {
```

Comment:
Let the `else` and initialize here the badger datastore instead of inside the `applyDefaults` function, because you are here in the specialized function to init the datastore whatever the case.